src/Native/System.Native/pal_networking.cpp

```diff
@@ -428,7 +429,112 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR
+#if !HAVE_GETHOSTBYNAME_R
+static int copy_hostent(struct hostent* from, struct hostent* to,
+                        char* buffer, size_t buflen)
+{
+    // FIXME: the implementation done for this function in https://github.com/dotnet/corefx/commit/6a99b74
+    //        requires testing when managed assemblies are built and tested on NetBSD. Until that time,
+    //        return an error code.
+    (void)from;   // unused arg
+    (void)to;     // unused arg
+    (void)buffer; // unused arg
+    (void)buflen; // unused arg
+    return -1;
+}
+
+/*
+Note: we're assuming that all access to these functions are going through these shims on the platforms, which do not provide
+      thread-safe functions to get host name or address. If that is not the case (which is very likely) race condition is
+      possible, for instance; if other libs (such as libcurl) call gethostby[name/addr] simultaneously.
+*/
+static pthread_mutex_t lock_hostbyx_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static int gethostbyname_r(char const* hostname, struct hostent* result,
+                           char* buffer, size_t buflen, hostent** entry, int* error)
+{
+    assert(hostname != nullptr);
+    assert(result != nullptr);
+    assert(buffer != nullptr);
+    assert(entry != nullptr);
+    assert(error != nullptr);
+
+    if (hostname == nullptr || entry == nullptr || error == nullptr || buffer == nullptr || result == nullptr)
+    {
+        return PAL_BAD_ARG;
+    }
+
+    pthread_mutex_lock(&lock_hostbyx_mutex);
+
+    *entry = gethostbyname(hostname);
+    if ((!(*entry)) || ((*entry)->h_addrtype != AF_INET) || ((*entry)->h_length != 4))
+    {
+         *error = h_errno;
+         *entry = NULL;
+    }
+    else
+    {
+         h_errno = copy_hostent(*entry, result, buffer, buflen);
+
+         if (h_errno != ERANGE)
+         {
+             *entry = result;
+         }
+         else
+         {
+             *entry = NULL;
+         }
```

Comment:
ERANGE is the only error we were returning from copy_hostent. I agree this could be made more general purpose by saying `h_errno == 0 ? result : nullptr`.